12章まで

cm-nakayasu-yuichi opened the pull request:
以下の理由で「式」を表す命名を `Expression` ではなく `Expressible`とした

- 書籍ではExpressionだが、NimbleのExpressionと名前衝突するため
- protocol として定義するので、Swift的な命名規則を採用した

cm-nakayasu-yuichi committed:
TODOを追加

cm-nakayasu-yuichi committed:
通貨の足し算テストを追記

cm-nakayasu-yuichi committed:
Moneyにplusメソッドを実装してテストをグリーンに

cm-nakayasu-yuichi committed:
テストにBankとExpressionの概念を追加。テストはレッド。
(ただし、Expressionという名前はNimbleと衝突することと、後に書籍ではインターフェイスとして定義されるため、書籍ではExpressionだがソースではExpressibleとした)

cm-nakayasu-yuichi committed:
BankとExpressibleファイルを追加

cm-nakayasu-yuichi committed:
Expressibleプロトコルを定義。Moneyに準拠させ、plusの戻り値の型をExpressibleとした

cm-nakayasu-yuichi committed:
Bankクラスを定義。reduceメソッドも定義してテストがグリーンになるまで仮実装した。

cm-TanakaKenji commented:
@cm-nakayasu-yuichi 
`Expressible` という命名よりも `Expression` とした方が良いと思いました。

理由は二つあります。

1. p.248に書かれているImposterの説明で、「なんらかの存在になりすますもの」として `Null Object` と `Composite`  を例に挙げられていました。 `Expression` はその役割に当てはまるので、そのままの命名で良いと思いました。

2. SwiftのProtocolの命名を考える上で以下の資料をよく参考にしています。この資料によるとSwiftのProtocolには三つのパターンがあり、それがなんであるかを表す `Is a` という使い方があります。(Collectionのようなもの)これが今回のImposterに当てはまるように思いました。
https://speakerdeck.com/gregheo/what-i-learned-from-55-star-swift-standard-library-protocols

cm-nakayasu-yuichi commented:
@cm-TanakaKenji 

名前衝突で怒られるので
```
MoneyTest.swift:56:22: 'Expression' is ambiguous for type lookup in this context
```

こういう書き方になります↓
```
let sum: tdd_by_example_swift.Expression = five.plus(five)
```

cm-TanakaKenji commented:
@cm-nakayasu-yuichi 
> 名前衝突

`tdd_by_example_swift.Expression` これ超気持ち悪いですねww
仕方ないのですが

cm-nakayasu-yuichi committed:
Expressibleよりもやはり書籍通りのExpressionの方がよいとのことで修正
https://github.com/cm-pliser-tdd-by-example/tdd-by-example-swift/pull/6#issuecomment-344286038

cm-nakayasu-yuichi committed:
Merge branch 'master' into chapter12

cm-nakayasu-yuichi commented:
@cm-TanakaKenji
Expressionに修正してみました。
またよろしくお願いします。
マージされ次第、13〜14章も書いていきます
